### PR TITLE
os current_exe using same approach as linux to get always the full ab…

### PR DESCRIPTION
**Pull request by devnexen**
…solute path

### CODE REVIEW

File: library/std/src/sys/unix/os.rs [hunk 1]
```diff
@@ -368,21 +368,12 @@ pub fn current_exe() -> io::Result<PathBuf> {
...
-            let path = PathBuf::from(<OsStr as OsStrExt>::from_bytes(filename));
-
-            // Prepend a current working directory to the path if
-            // it doesn't contain an absolute pathname.
-            if filename[0] == b'/' { Ok(path) } else { getcwd().map(|cwd| cwd.join(path)) }
-        }
+    match crate::fs::read_link("/proc/self/path/a.out") {
+        Err(ref e) if e.kind() == io::ErrorKind::NotFound => Err(io::Error::new_const(
+            io::ErrorKind::Uncategorized,
+            &"no /proc/self/a.out available. Is /proc mounted?",
```
Comment (bjorn3) [inline library/std/src/sys/unix/os.rs:374]: Maybe still use `getexecname` if `/proc` is not mounted? I am guessing this could happen inside containers.

Comment (devnexen) [reply library/std/src/sys/unix/os.rs:374]: fair point

**Comment by Mark-Simulacrum** [comment]
The commit says "same approach as linux", but I am not seeing similar code elsewhere in this file. Regardless, this is not really my area of expertise so.. let's try r? @joshtriplett perhaps?

**Comment by devnexen** [comment]
originally it was closer but difference now there is fallback code path.

**Comment by joshtriplett** [comment]
I don't know Solaris, but the approach seems reasonable, and I see various references confirming that `/proc/self/path/a.out` works on Solaris.

@bors r+

**Pull request merged**